### DISCUSSION TITLE
Loading hostfxr.dll fails (apparently)

### DISCUSSION THREAD

**Discussion by tvidvei** [dotnet/runtime#79391] [Q&A] [answered]
I use Excel-DNA 1.6.1-beta1 in an Excel Add-In based on .NET 6.0.11.   The Add-In is published as a FDD (Framework Dependent Deployment) application. It works fine on my machine which (of course) has .NET 6.0 Sdk installed.  But on a users machine I get this error:
![image](https://user-images.githubusercontent.com/5440347/206419661-7177e26c-b66c-4fc2-bb04-beb54f25d94b.png)
On the users machine,  .NET Desktop Runtime 6.0.11is installed.  The command "dotnet --info" returns what I thought confirms that the .NET installation is ok:
![image](https://user-images.githubusercontent.com/5440347/206420198-fdc8af92-8a19-4255-ae99-91b4645aca98.png)
I thought that .NET Desktop Runtime should be sufficient for having the Add-In to run properly.  But some places I find statements that the .NET6.0 SDK has to be installed on the client machines to have a .NET60-based FDD Application to run.  Eg. here: https://www.thecodebuzz.com/self-contained-vs-framework-dependent-deployment-hostfxr-fdd-vs-scd/.  That sounds strange to me.  Why do we have runtimes if the SDK is needen in any case?

My question:  Any clue on what the error could be, and how to fix it?

**Comment by vitek-karas** [comment, marked as answer]
Is the client machine running 32bit Excel/plugin? If so, then it is probably looking for x86 runtime installation which is probably not there.

The code which causes the error seems to be here: https://github.com/Excel-DNA/ExcelDna/blob/18fe0ad01e779e6fccb579aed78399ce495116cf/Source/ExcelDna.Host/host.cpp

As far as I can tell it's using nethost, so you should be able to get more detailed information by turning on host tracing:
`COREHOST_TRACE=1` and `COREHOST_TRACEFILE=C:\temp\host.txt`. 
After the repro the `host.txt` should contain some details on what it was looking for and where.

FDD application only needs the runtime - not the SDK. That said, SDK contains the runtime, so if there's SDK on the machine then it will work as well. I don't know why the above mentioned blog says that you need SDK - that was never the case.

**Comment by tvidvei** [comment]
Thanks a lot for your answer!

Yes, the client machine, and also my own machine where the development takes place, runs 32bit Excel, while the machines themselves are 64bit with Windows 10.

I will do what you propose and report back!

Regards,
Tor

**Reply by vitek-karas** [reply]
The 32bit Excel will likely require a 32bit runtime install. It's also likely that your developer machine already has a 32bit runtime (and SDK) installed (VS does that I think), but the client machine won't have it unless installed by something or manually.

**Reply by tvidvei** [reply]
Nail on!  Installing 32bit runtime, as you proposed, solved the problems. I was totally blindfolded by the fact that our machines are 64bit. Since the add-in is compiled for "Any CPU", I had no attention on 32bit issue with Excel.

Again: Thanks a lot for your reply which solved our problem!

**Reply by vitek-karas** [reply]
> Since the add-in is compiled for "Any CPU"

Custom hosting of the runtime can't be "Any CPU" - it's native code, so it has to be compiled for a specific architecture. In the case of an in-proc plugin (like Excel) it needs to match the architecture of the process. And the custom host can only load runtime of the matching architecture into the process (same rules). So even if the managed parts of the component are compiled as "Any CPU", the hosting code is architecture specific.

**Reply by tvidvei** [reply]
I understand.  It has been a steep learning curve as the add-in was migrated from .NET Framework just recently.  And most of our developent on NET Core and successors are to be used in various services hosted on Azure, and are not to be deployed as desktop applications. Having a proper knowledge on the platform architectures, hosting, and deployment alternatives would have saved us a lot of time and frustration over the last couple of days.  But now we are at least better equipped. Thanks again!

**Comment by tvidvei** [comment]
That's right, my machine has both 32bit 64bit SDKs installed.  We have already done a test on one machine which was promising.  So I guess what you proposed is the solution.  I'll send an update when we have tested it on the machine we had the problems on.
